frame/name-service/src/lib.rs

```diff
@@ -0,0 +1,480 @@
+		_,
+		Blake2_128Concat,
+		NameHash,
+		Registration<T::AccountId, BalanceOf<T>, T::BlockNumber>,
+	>;
+
+	/// This resolver maps name hashes to an account
+	#[pallet::storage]
+	#[pallet::getter(fn resolve)]
+	pub(super) type Resolvers<T: Config> = StorageMap<_, Blake2_128Concat, NameHash, T::AccountId>;
+
+	// Your Pallet's events.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		// TODO: Potentially make events more lightweight
+		/// A new `Commitment` has taken place.
+		Committed { sender: T::AccountId, who: T::AccountId, hash: CommitmentHash },
+		/// A new `Registration` has taken added.
+		NameRegistered { name_hash: NameHash, owner: T::AccountId, expiry: T::BlockNumber },
+		/// A `Registration` has been transferred to a new owner.
+		Transfer { from: T::AccountId, to: T::AccountId },
+		/// A `Registration` has been renewed.
+		NameRenewed { name_hash: NameHash, expires: T::BlockNumber },
+		/// An address has been set for a name hash to resolve to.
+		AddressSet { name_hash: NameHash, address: T::AccountId },
+		/// An address was deregistered.
+		AddressDeregistered { name_hash: NameHash },
+	}
+
+	// Your Pallet's error messages.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// This commitment hash already exists in storage.
+		AlreadyCommitted,
+		/// It has not passed the minimum waiting period to reveal a commitement.
+		TooEarlyToReveal,
+		/// The commitment cannot yet be removed. Has not expired.
+		CommitmentNotExpired,
+		/// The commitment has expired.
+		CommitmentExpired,
+		/// This commitment does not exist.
+		CommitmentNotFound,
+		/// This name is already registered.
+		AlreadyRegistered,
+		/// This registration does not exist.
+		RegistrationNotFound,
+		/// The sender is not the registration owner.
+		NotRegistrationOwner,
+		/// This resolver does not exist.
+		ResolverNotFound,
+		/// This registration has expired.
+		RegistrationExpired,
+		/// This registration has not yet expired.
+		RegistrationNotExpired,
+		/// Conversion error
+		ConversionError,
+	}
+
+	// Your Pallet's callable functions.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		// TODO: Should we allow registration on behalf of?
+		#[pallet::weight(0)]
+		pub fn commit(
+			origin: OriginFor<T>,
+			who: T::AccountId,
+			commitment_hash: CommitmentHash,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			ensure!(!Commitments::<T>::contains_key(commitment_hash), Error::<T>::AlreadyCommitted);
+
+			let block_number = frame_system::Pallet::<T>::block_number();
+			let deposit = T::CommitmentDeposit::get();
+
+			T::Currency::reserve(&sender, deposit)?;
+
+			let commitment = Commitment { who: who.clone(), when: block_number, deposit };
+
+			Commitments::<T>::insert(commitment_hash, commitment);
+			Self::deposit_event(Event::<T>::Committed { sender, who, hash: commitment_hash });
+
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn reveal(
```

Comment:
needs to be `#[tranasctional]`

Reply:
It is better to follow our verify first, change last pattern.  
Here the problem is that `do_register` can fail, after the storage was modified.  
This then leaves the pallet in an unknown state.

Reply:
Verify first, change last is a bad pattern that exists only because Substrate was very incapable.
There is almost no reason to continue use this pattern with `tranasctional`.

Reply:
generally I agree with this statement, as long as there is no ability to nest transactional layers, then it should be fine.